src/parser/core/Module.ts

```diff
@@ -0,0 +1,273 @@
+import {cloneDeep} from 'lodash'
+import 'reflect-metadata'
+
+import {Event, Ability} from 'fflogs'
+import Parser from './Parser'
+
+export const DISPLAY_ORDER = {
+	TOP: 0,
+	DEFAULT: 50,
+	BOTTOM: 100,
+}
+
+export function dependency(target: Module, prop: string) {
+	const dependency = Reflect.getMetadata('design:type', target, prop)
+	const constructor = target.constructor as typeof Module
+
+	// Make sure we're not modifying every single module
+	if (constructor.dependencies === Module.dependencies) {
+		constructor.dependencies = []
+	}
+
+	// If the dep is Object, it's _probably_ from a JS file. Fall back to simple handling
+	if (dependency === Object) {
+		constructor.dependencies.push(prop)
+		return
+	}
+
+	// Check that the dep is actually a module
+	if (!Module.isPrototypeOf(dependency)) {
+		throw new Error(`${constructor.name}'s dependency \`${prop}\` is invalid. Expected \`Module\`, got \`${dependency.name}\`.`)
+	}
+
+	constructor.dependencies.push({
+		handle: dependency.handle,
+		prop,
+	})
+}
+
+interface MappedDependency {
+	handle: string
+	prop: string
+}
+
+type Filter<T extends Event> = Partial<T> & Partial<{
+	abilityId: Ability['guid']
+	to: 'player' | 'pet' | T['targetID']
+	by: 'player' | 'pet' | T['sourceID']
+}>
+
+type HookCallback<T extends Event> = (event: T) => void
+
+interface Hook<T extends Event> {
+	events: T['type'][]
+	filter: Filter<T>
+	callback: HookCallback<T>
+}
+
+export default class Module {
+	static dependencies: Array<string | MappedDependency> = []
+	static displayOrder = DISPLAY_ORDER.DEFAULT
+
+	private static _handle: string
+	static get handle() {
+		if (!this._handle) {
+			const handle = this.name.charAt(0).toLowerCase() + this.name.slice(1)
+			console.error(`\`${this.name}\` does not have an explicitly set handle. Using \`${handle}\`. This WILL break in minified builds.`)
+			this._handle = handle
+		}
+		return this._handle
+	}
+	static set handle(value) {
+		this._handle = value
+	}
```

Comment:
Just a compatibility note: in the current ES class properties spec, you can't just write `static handle = 'foo'` in a class body; that won't invoke setters and will just define a new own property that bypasses the setter. You'd need to do a `Foo.handle = 'foo'` after the class is defined.

This currently only works because `babel-preset-react-app` is forcing babel to transform class fields in the old spec mode: https://github.com/facebook/create-react-app/blob/master/packages/babel-preset-react-app/create.js#L115-L122

Fortunately, these setters don't rely on causing side effects so it'd probably be safe even if transforming in spec mode.

Reply:
That's unfortunate, a lot of stuff around the codebase uses that pattern.

I doubt CRA'll be changing that in the short term, so should be OK for now, but worth keeping in mind.